File: src/stackqueue/checkout/Checkout.java
package stackqueue.checkout;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.LinkedList;
import java.util.Queue;

/**
 * ClassName:Checkout
 * Package:stackqueue.checkout
 * Description:
 *
 * @date:2024/12/27 14:56
 * @author: Junquan Yi
 */
class Checkout {
    // 题目: LCR 184. 设计自助结算系统,middle
    // https://leetcode.cn/problems/dui-lie-de-zui-da-zhi-lcof/description/
    Queue<Integer> queue; // 负责正常的插入删除
    Deque<Integer> deque; // 负责维护单调队列
    public Checkout() {
        queue = new LinkedList<>();
        deque = new LinkedList<>();
    }

    public int get_max() {
        if(deque.isEmpty()) return -1;
        return deque.peekFirst();
    }

    public void add(int value) {
        queue.offer(value);
        while(!deque.isEmpty() && deque.peekLast() < value) {
            deque.pollLast();
        }
        deque.offerLast(value);
    }

    public int remove() {
        if(queue.isEmpty()) return -1;
        int ans = queue.poll();
        if(ans==deque.peekFirst()){
            deque.pollFirst();
        }
        return ans;
    }
}

/**
 * Your Checkout object will be instantiated and called as such:
 * Checkout obj = new Checkout();
 * int param_1 = obj.get_max();
 * obj.add(value);
 * int param_3 = obj.remove();
 */

File: python/stackqueue/dandiaoqueue/Checkout.py
from collections import deque

class Checkout:

    def __init__(self):
        self.q1 = deque() # 普通队列
        self.q2 = deque() # 单调队列

    def get_max(self) -> int:
        if not self.q1:
            return -1
        return self.q2[0]

    def add(self, value: int) -> None:
        self.q1.append(value)
        while self.q2 and value >= self.q2[-1]:
            self.q2.pop()
        self.q2.append(value)

    def remove(self) -> int:
        if not self.q1:
            return -1
        ans = self.q1.popleft() # 队首剔除
        if ans == self.q2[0]:
            self.q2.popleft()
        return ans

        


# Your Checkout object will be instantiated and called as such:
# obj = Checkout()
# param_1 = obj.get_max()
# obj.add(value)
# param_3 = obj.remove()